test/unit/storage/validators/QuotaValidator.test.ts

```diff
@@ -0,0 +1,120 @@
+import type { QuotaStrategy } from '../../../../src/storage/quota/QuotaStrategy';
+import type { SizeReporter } from '../../../../src/storage/size-reporter/SizeReporter';
+import { QuotaValidator } from '../../../../src/storage/validators/QuotaValidator';
+import { guardStream } from '../../../../src/util/GuardedStream';
+import type { Guarded } from '../../../../src/util/GuardedStream';
+import { guardedStreamFrom, readableToString } from '../../../../src/util/StreamUtil';
+
+describe('QuotaValidator', (): void => {
+  let mockedStrategy: jest.Mocked<QuotaStrategy>;
+  let validator: QuotaValidator;
+  let identifier: ResourceIdentifier;
+  let mockMetadata: RepresentationMetadata;
+  let mockData: Guarded<Readable>;
+  let mockInput: ValidatorInput;
+  let mockReporter: jest.Mocked<SizeReporter<any>>;
+
+  beforeEach((): void => {
+    jest.clearAllMocks();
+    identifier = { path: 'http://localhost/' };
+    mockMetadata = new RepresentationMetadata();
+    mockData = guardedStreamFrom([ 'test string' ]);
+    mockInput = {
+      representation: new BasicRepresentation(mockData, mockMetadata),
+      identifier,
+    };
+    mockReporter = {
+      getSize: jest.fn(),
+      getUnit: jest.fn(),
+      calculateChunkSize: jest.fn(),
+      estimateSize: jest.fn().mockResolvedValue(8),
+    };
+    mockedStrategy = {
+      reporter: mockReporter,
+      limit: { unit: 'bytes', amount: 8 },
+      getAvailableSpace: jest.fn().mockResolvedValue({ unit: 'bytes', amount: 10 }),
+      estimateSize: jest.fn().mockResolvedValue({ unit: 'bytes', amount: 8 }),
+      createQuotaGuard: jest.fn().mockResolvedValue(guardStream(new PassThrough())),
+    };
+    validator = new QuotaValidator(mockedStrategy);
+  });
+
+  describe('handle()', (): void => {
+    // Step 2
+    it('should destroy the stream when estimated size is larger than the available size.', async(): Promise<void> => {
+      mockedStrategy.estimateSize.mockResolvedValueOnce({ unit: 'bytes', amount: 11 });
+
+      const result = validator.handle(mockInput);
+      await expect(result).resolves.toBeDefined();
+      const awaitedResult = await result;
+
+      const prom = new Promise<void>((resolve, reject): void => {
+        awaitedResult.data.on('error', (): void => resolve());
+        awaitedResult.data.on('end', (): void => reject(new Error('reject')));
+      });
+
+      // Consume the stream
+      await expect(readableToString(awaitedResult.data))
+        .rejects.toThrow('Quota exceeded: Advertised Content-Length is');
+      await expect(prom).resolves.toBeUndefined();
+    });
+
+    // Step 3
+    it('should destroy the stream when quota is exceeded during write.', async(): Promise<void> => {
+      const createQuotaGuardSpy = jest.spyOn(mockedStrategy, 'createQuotaGuard')
+        .mockResolvedValueOnce(guardStream(new PassThrough({
+          async transform(this): Promise<void> {
+            this.destroy();
+          },
+        })));
```

Comment:
Similarly, since these are actually jest functions you also don't need a spy for this

```suggestion
      mockedStrategy.createQuotaGuard.mockResolvedValueOnce(guardStream(new PassThrough({
          async transform(this): Promise<void> {
            this.destroy();
          },
        })));
```